### DISCUSSION TITLE
WebGPU copyTextureToBuffer auto convert YUV to RGB

### DISCUSSION THREAD

**Discussion by huotong1212** [gpuweb/gpuweb#4395] [Q&A] [answered]
my input videoframe is like this:
![1701005036043](https://github.com/gpuweb/gpuweb/assets/31182767/e9bbdb41-8075-477a-9cff-f9928b974f21)
it' NV12 format , Y channel size 1280 * 720 , UV channel size is 1280 * 360

create texture like this:
```typescripts
        const textureDescriptor: GPUTextureDescriptor = {
            size: {
                width: this.width,  // 1280 
                height: this.height,  // 720
            },
            format: "r8unorm",
            usage: GPUTextureUsage.TEXTURE_BINDING | GPUTextureUsage.COPY_DST | GPUTextureUsage.COPY_SRC | GPUTextureUsage.RENDER_ATTACHMENT
        }

        this.texture = this.device.createTexture(textureDescriptor);
        
        this.device.queue.copyExternalImageToTexture(
            {source: videoFrame as any},
            {texture: this.texture},
            {
                width: this.width,
                height: this.height
            }
        )
        
        const bytesPerRow = this.width * this.channels;
        commandEncoder.copyTextureToBuffer( // to show the texture data
            {texture: this.texture},
            {
                buffer: this.videoFrameBuffer,
                offset: 0, bytesPerRow: bytesPerRow, rowsPerImage: this.height
            }, {
                width: this.width, height: this.height, depthOrArrayLayers: 1
            }
        );
```
i just want texture ( format r8unorm) only accept the y channel and uv channel and dont change the data,but i got texture data like this 
![1701005291283](https://github.com/gpuweb/gpuweb/assets/31182767/29490866-7783-42bd-a9e0-1fc8deeb70ad)

as we can see, the Y 162,162,162,163 become 168,168,168,169. i am confused about this, how can i directly the raw data 162,162,162,163

does anyone know the reason?,I will be extremely grateful for any help.

**Comment by Kangz** [comment, marked as answer]
The call `copyExternalImageToTexture` always copy to RGBA data. This means that there is an implicit YUV->RGBA conversion happening in your code.

At this time there is no way to access the Y and YUV planes for a `VideoFrame` directly, except by manually uploading their contents to a texture (using `writeTexture`, or `copyBufferToTexture`)

**Reply by Andrea3000** [reply]
Hi @Kangz,

I'm also trying to access individual planes of a YUV VideoFrame.
You had suggested to use copyBufferToTexture. Does it mean that I can copy the Y, U and V planes to 3 separate GPUTextures without copying the data back to the CPU?
If so, how? It isn't clear to me how this would be possible?

Thanks

**Reply by Kangz** [reply]
My suggestion was to roundtrip through the CPU. https://github.com/gpuweb/gpuweb/issues/5281 has discussions about sampling the Y and UV directly in the shaders but no concrete progress.
